backend/src/apiserver/resource/resource_manager.go

```diff
@@ -195,616 +203,989 @@ func (r *ResourceManager) ArchiveExperiment(ctx context.Context, experimentId st
+// Manifest's namespace gets overwritten with the run.Namespace.
+// Creating a run from recurring run prioritizes recurring run's pipeline spec over the run's one.
+func (r *ResourceManager) CreateRun(ctx context.Context, run *model.Run) (*model.Run, error) {
+	expNs, expId, err := r.validateExperimentNamespace(run.Namespace, run.ExperimentId)
 	if err != nil {
-		return nil, err
+		return nil, util.Wrapf(err, "Failed to create a run. Specify a valid experiment id and namespace combination")
 	}
+	run.ExperimentId = expId
+	run.Namespace = expNs
 
-	// Generate Run Id and timestamps.
-	uuid, err := r.uuid.NewRandom()
+	// Fetch pipeline version from pipeline spec
+	var wfTemplate *template.Template
+	pipelineVersion, err := r.fetchPipelineVersionFromPipelineSpec(run.PipelineSpec, run.DisplayName, run.Description, run.Namespace)
 	if err != nil {
-		return nil, util.NewInternalServerError(err, "Failed to generate run ID.")
+		return nil, util.Wrapf(err, "Failed to create a run. Specify a valid pipeline spec")
 	}
-	runId := uuid.String()
-	runAt := r.time.Now().Unix()
-
-	// Create template for this run. Template can be argo template or IR. New templates may be added in the future.
-	tmpl, err := template.New(manifestBytes)
-	if err != nil {
-		return nil, err
+	manifest := run.PipelineSpec.PipelineSpecManifest
+	if manifest == "" {
+		manifest = run.PipelineSpec.WorkflowSpecManifest
 	}
-	runWorkflowOptions := template.RunWorkflowOptions{
-		RunId: runId,
-		RunAt: runAt,
+	if manifest == "" && pipelineVersion != nil {
+		manifest = pipelineVersion.PipelineSpec
 	}
-
-	// Convert apiRun to model Run.
-	modelRunDetail, err := r.ToModelRunDetail(apiRunInterface, runId, runAt, string(manifestBytes), tmpl.GetTemplateType())
-	if err != nil {
-		return nil, util.Wrap(err, "Error creating model RunDetail")
+	if pipelineVersion == nil {
+		pipelineVersion, wfTemplate, err = r.createPipelineFromSpecIfNoExisting(manifest, run.Namespace, run.DisplayName, run.Description)
+		if err != nil {
+			return nil, util.Wrap(err, "Failed to create a run due to error fetching pipeline version from pipeline spec")
+		}
 	}
-
-	// Convert modelRun into execution spec.
-	executionSpec, err := tmpl.RunWorkflow(&modelRunDetail.Run, runWorkflowOptions)
+	run.PipelineSpec.PipelineId = pipelineVersion.PipelineId
+	run.PipelineSpec.PipelineVersionId = pipelineVersion.UUID
+	run.PipelineSpec.PipelineName = pipelineVersion.Name
+	// Get manifest from either of the two places:
+	// (1) raw manifest in pipeline_spec
+	// (2) pipeline version in resource_references
+	// And the latter takes priority over the former when the manifest is from pipeline_spec.pipeline_id
+	// workflow/pipeline manifest and pipeline id/version will not exist at the same time, guaranteed by the validation phase
+	var tmpl template.Template
+	if wfTemplate == nil {
+		tempBytes, err2 := r.fetchTemplateFromPipelineVersionId(run.PipelineSpec.PipelineVersionId)
+		if err2 != nil {
```

Comment:
nit: seems like unnecessary to name this as `err2` instead of `err`?

Reply:
Thank you. Fixed here and in `CreateJob` function.